traceevents/src/main/java/com/tomtom/kotlin/traceevents/TraceEventConsumerCollection.kt

```diff
@@ -32,33 +28,45 @@ import kotlin.reflect.jvm.javaMethod
  */
 class TraceEventConsumerCollection {
 
-    fun add(traceEventConsumer: TraceEventConsumer) {
-        traceEventsConsumers.add(traceEventConsumer)
+    fun add(traceEventConsumer: TraceEventConsumer, contextRegex: Regex?) {
+        traceEventsConsumersWithContext.add(TraceEventConsumerWithContext(traceEventConsumer, contextRegex))
     }
 
-    fun remove(traceEventConsumer: TraceEventConsumer) {
-        traceEventsConsumers.remove(traceEventConsumer)
+    fun remove(traceEventConsumer: TraceEventConsumer, contextRegex: Regex? = null) {
+        for (consumer in traceEventsConsumersWithContext.asIterable()
+            .filter {
+                it.traceEventConsumer == traceEventConsumer &&
+                        (contextRegex == null || it.contextRegex == contextRegex)
+            }) {
+            traceEventsConsumersWithContext.remove(consumer)
+        }
     }
 
-    fun all(): Iterable<TraceEventConsumer> {
-        return traceEventsConsumers.asIterable()
+    fun all(contextRegex: Regex? = null): Iterable<TraceEventConsumer> {
+        return traceEventsConsumersWithContext.asIterable()
+            .filter { contextRegex == null || it.contextRegex == contextRegex }
+            .map { it.traceEventConsumer }
     }
 
     suspend fun consumeTraceEvent(traceEvent: TraceEvent) {
 
         // Trace events are offered to every handler that can handle the event.
-        for (traceEventConsumer in traceEventsConsumers) {
+        for (traceEventConsumerWithContext in traceEventsConsumersWithContext) {
 
             /**
              * Check what type of trace event consumer we're dealing with. There are two types of
              * consumer: generic trace event handlers, that will be offered all events always, and
              * [TraceEventListener]-derived trace event consumer, that can only deal with the events they
              * define as implemented functions.
              */
-            if (traceEventConsumer is GenericTraceEventConsumer) {
-                traceEventConsumer.consumeTraceEvent(traceEvent)
-            } else {
-                handleSpecificConsumer(traceEventConsumer, traceEvent)
+            if (traceEventConsumerWithContext.contextRegex == null ||
+                traceEventConsumerWithContext.contextRegex.matches(traceEvent.context)
+            ) {
+                if (traceEventConsumerWithContext.traceEventConsumer is GenericTraceEventConsumer) {
+                    traceEventConsumerWithContext.traceEventConsumer.consumeTraceEvent(traceEvent)
```

Comment:
Given that one `TraceEventConsumer` instance can be registered multiple times with multiple (overlapping) regexes, it can also be called multiple times here which seems logical wrong to me.

```
// GIVEN a trace event consumer is registered two times with overlapping regexes
// WHEN a trace event is consumed that matches both regexes.
// THEN the trace event consumer is only invoked once.
```

An alternative, an much simpler is to enforce that a trace event consumer can only be registered once. That would also simply the `TraceEventConsumerCollection.remove()` function as in that case the regex argument is not needed any longer. As the client doesn't keep track of traceevent consumer instances and regexes any longer, I would prefer this option:

```
// WHEN a trace event consumer is registered when it is already registered 
// THEN fail with IllegalArgumentException.
```

Reply:
Why would you not allow multiple registrations of a single consumer with different regex's?
I agree a consumer should be called only once, but it seems to fine to register it multiple times.

Reply:
Ok. Yeah, indeed. I misread the implementation of remove(). Lets change this to only call a consumer once here.